### ISSUE TITLE
Move to log4j 2.x?

### ISSUE THREAD

**Issue by colindean** [Log4s/log4s#13]
log4s depends on slf4j-api 1.7.12, which depends on log4j 1.2.17.

```
+--- org.log4s:log4s_2.11:1.2.1
|    +--- org.scala-lang:scala-library:2.11.7 -> 2.11.8
|    \--- org.slf4j:slf4j-api:1.7.12
+--- org.slf4j:slf4j-log4j12:1.7.12
|    +--- org.slf4j:slf4j-api:1.7.12
|    \--- log4j:log4j:1.2.17
```

I'm just starting to investigate the benefits of log4j 2.x, but figure it's worth investigation because log4j 1.2 is EOL'd and I've got this cognitive bias against EOL'd things in my production code...

The latest slf4j is 1.7.12, but seems like it can only use log4j 1.2.17.

Maybe I can use something else instead of log4j as the backend for slf4j? [AVSL](http://software.clapper.org/avsl/) seems friendly, but creates some logistics for us to migrate (technically a breaking change for our on-premises product, new release required, yada yada).

I'm interested in what others using log4s are doing.

**Comment by larsrh** [comment]
I don't understand where this is supposed to come from. I don't see any non-test dependencies on `log4j` in this build. So you should be fine in using the [sl4j binding for log4j](https://logging.apache.org/log4j/2.0/log4j-slf4j-impl/index.html).

**Comment by colindean** [comment]
Isn't this it [here](https://github.com/Log4s/log4s/blob/master/project/Build.scala#L270)?

Although, in my dependencies, I've specifically depended on `org.slf4j:slf4j-log4j12:1.7.12`. I'm recalling now that "things didn't work" absent that, and that maybe newer versions of slf4j-log4j provide the same interface, so it just works. I've never tried just using a newer version.

Interestingly enough, the question of logging frameworks came up on [/r/scala](https://www.reddit.com/r/scala). In [this comment thread](https://www.reddit.com/r/scala/comments/537imy/is_there_a_purely_functional_and_async_logging/d7qt6qe?context=3), I asked basically the same question here. The reply to my question suggested not to rely on transitive dependencies.

**Comment by larsrh** [comment]
No, this is a dependency `slf4j-api`. That's just the API, carrying no dependency on any logging framework at all. By extension, `log4s` doesn't depend on any logging framework. If in your code, you only depend on `log4s`, but not on `log4j` or anything similar, you will get warnings from `slf4j-api` that it couldn't find any binding. The dependency tree you posted in the original issue is _not_ the tree from `log4s`, so you must have had an explicit dependency on `log4j` somewhere. I suggest replacing that one by `org.apache.logging.log4j:log4j-slf4j-impl:2.6.2`; according to the [log4j site](https://logging.apache.org/log4j/2.x/log4j-slf4j-impl/dependencies.html#Dependency_Tree) that's the latest one.

**Comment by colindean** [comment]
I think I understand better now. `log4j-slf4j-impl` implements the slf4j interface, but then needs `log4j-core` to do the real work.

```diff
diff --git a/build.gradle b/build.gradle
index 377a5b1..45ac557 100644
--- a/build.gradle
+++ b/build.gradle
@@ -100,7 +100,8 @@ dependencies {

     // logging
     compile 'org.log4s:log4s_2.11:1.2.1'
-    compile 'org.slf4j:slf4j-log4j12:1.7.12'
+    compile 'org.apache.logging.log4j:log4j-slf4j-impl:2.6.2'
+    compile 'org.apache.logging.log4j:log4j-core:2.6.2'

     //internationalization
     scalaCompilerPlugin 'tv.cntt:xgettext_2.11:1.3'
```

is what I'll do eventually.

Maybe the log4s documentation could be improved to clarify that this, or something like this, is what is needed? That's not really straightforward.

**Comment by sarahgerweck** [comment]
I've just discovered that apparently I'm not getting email notifications for this project, so I apologize for the radio silence. As previously discussed, I believe that SLF4J bindings should allow you to use basically any framework you want. I agree that some better documentation would be useful here for people who aren't familiar with SLF4J. (Some of us have been working in the Java ecosystem for a long time and forget that not everybody went on the long, messy journey that was Java logging before everybody more or less agreed on SLF4J as a standard façade on top of a few major options.)

**Comment by sarahgerweck** [comment]
I've opened #19 to track the documentation improvements that are needed here. As I said before, I do agree that it would be nice to have a getting-started guide that covers this.
